crates/bevy_ecs/src/event/trigger.rs

```diff
@@ -0,0 +1,255 @@
+use crate::{
+    component::ComponentId,
+    entity::Entity,
+    event::{EntityEvent, Event},
+    observer::{CachedObservers, TriggerContext},
+    traversal::Traversal,
+    world::DeferredWorld,
+};
+use bevy_ptr::PtrMut;
+use core::marker::PhantomData;
+
+pub trait Trigger<E: Event> {
+    fn trigger(
+        &mut self,
+        world: DeferredWorld,
+        observers: &CachedObservers,
+        trigger_context: &TriggerContext,
+        event: &mut E,
+    );
+}
+
+#[derive(Default)]
+pub struct GlobalTrigger;
+
+impl<E: Event> Trigger<E> for GlobalTrigger {
+    fn trigger(
+        &mut self,
+        world: DeferredWorld,
+        observers: &CachedObservers,
+        trigger_context: &TriggerContext,
+        event: &mut E,
+    ) {
+        self.trigger_internal(world, observers, trigger_context, event.into());
+    }
+}
+
+impl GlobalTrigger {
+    fn trigger_internal(
+        &mut self,
+        mut world: DeferredWorld,
+        observers: &CachedObservers,
+        trigger_context: &TriggerContext,
+        mut event: PtrMut,
+    ) {
+        // SAFETY: there are no outstanding world references
+        unsafe {
+            world.as_unsafe_world_cell().increment_trigger_id();
+        }
```

Comment:
Nit: technically here `observers` is a reference to data in `world`. It doesn't really matter for calling `.increment_trigger_id()`, but it might confuse someone that might want to update this code in the future and expect no reference to the `world` to actually exist.

Reply:
Good catch!